### PR TITLE
Batch up requests with more than 10 target group changes

### LINKED ISSUE #256: [aws_autoscaling_group] Attaching/Detaching a lot of loadbalancers fails
_This issue was originally opened by @freeman as hashicorp/terraform#8718. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hello,

We heavily use modules in our terraform stacks.
When updating some core modules this resulted in massive updates to the load balancers attached to some of our autoscaling groups and we are getting this error:

```
* aws_autoscaling_group.web: [WARN] Error updating Load Balancers for AutoScaling Group (DEV-AS-XXX-WEB), error: ValidationError: Trying to update too many Load Balancers/Target Groups at once. The limit is 10
    status code: 400, request id: xxx3fc5c-751b-11e6-8d48-bf41d1251xxx
```
### Terraform Version

Using terraform v0.7.3
### Affected Resource(s)
- aws_autoscaling_group
### Expected Behavior

The update should have succeded
### Actual Behavior

We got the error mentionned earlier
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. Create an infrastructure with more than 10 load balancers attached to an autoscaling group
2. `terraform apply`
3. You will get the error mentionned

From looking at the source this happens [here](https://github.com/hashicorp/terraform/blob/v0.7.3/builtin/providers/aws/resource_aws_autoscaling_group.go#L539-L557)

According to [Auto Scaling Limits](https://docs.aws.amazon.com/autoscaling/latest/userguide/as-account-limits.html) this will also be an issue with target groups.

I guess the feature request is for terraform to proceed in batches of 10 modifications at a time.

### PR DESCRIPTION
**Pull request by beezly**
The AWS API only supports adding/removing 10 requests at a time. 

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #256 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Batch up requests for Attach/DetachLoadBalancerTargetGroups into groups of 10
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: aws/resource_aws_autoscaling_group.go [hunk 1]
```diff
@@ -920,22 +920,47 @@ func resourceAwsAutoscalingGroupUpdate(d *schema.ResourceData, meta interface{})
...
+			batches = append(batches, remove)
+
+			for _, batch := range batches {
+				_, err := conn.DetachLoadBalancerTargetGroups(&autoscaling.DetachLoadBalancerTargetGroupsInput{
+					AutoScalingGroupName: aws.String(d.Id()),
+					TargetGroupARNs:      batch,
+				})
+				if err != nil {
+					return fmt.Errorf("Error updating Load Balancers Target Groups for AutoScaling Group (%s), error: %s", d.Id(), err)
+				}
```
Comment (bflad) [inline aws/resource_aws_autoscaling_group.go]: We need to ensure that every batch becomes fully applied before moving to the next 👍 

```suggestion
				}

				if err := waitUntilAutoscalingGroupLoadBalancerTargetGroupsRemoved(conn, d.Id()); err != nil {
					return fmt.Errorf("error describing AutoScaling Group (%s) Load Balancer Target Groups being removed: %s", d.Id(), err)
				}
```

Definition for new `waitUntilAutoscalingGroupLoadBalancerTargetGroupsRemoved` function:

```go
func waitUntilAutoscalingGroupLoadBalancerTargetGroupsRemoved(conn *autoscaling.AutoScaling, asgName string) error {
	input := &autoscaling.DescribeLoadBalancerTargetGroupsInput{
		AutoScalingGroupName: aws.String(asgName),
	}
	var tgRemoving bool

	for {
		output, err := conn.DescribeLoadBalancerTargetGroups(input)

		if err != nil {
			return err
		}

		for _, tg := range output.LoadBalancerTargetGroups {
			if aws.StringValue(tg.State) == "Removing" {
				tgRemoving = true
				break
			}
		}

		if tgRemoving {
			tgRemoving = false
			input.NextToken = nil
			continue
		}

		if aws.StringValue(output.NextToken) == "" {
			break
		}

		input.NextToken = output.NextToken
	}

	return nil
}
```

File: aws/resource_aws_autoscaling_group.go [hunk 2]
```diff
@@ -920,22 +920,47 @@ func resourceAwsAutoscalingGroupUpdate(d *schema.ResourceData, meta interface{})
...
+
+			for _, batch := range batches {
+				_, err := conn.AttachLoadBalancerTargetGroups(&autoscaling.AttachLoadBalancerTargetGroupsInput{
+					AutoScalingGroupName: aws.String(d.Id()),
+					TargetGroupARNs:      batch,
+				})
+
+				if err != nil {
+					return fmt.Errorf("Error updating Load Balancers Target Groups for AutoScaling Group (%s), error: %s", d.Id(), err)
+				}
```
Comment (bflad) [inline aws/resource_aws_autoscaling_group.go]: We need to ensure that every batch becomes fully applied before moving to the next 👍 

```suggestion
				}

				if err := waitUntilAutoscalingGroupLoadBalancerTargetGroupsAdded(conn, d.Id()); err != nil {
					return fmt.Errorf("error describing AutoScaling Group (%s) Load Balancer Target Groups being added: %s", d.Id(), err)
				}
```

Definition for new `waitUntilAutoscalingGroupLoadBalancerTargetGroupsAdded` function:

```go
func waitUntilAutoscalingGroupLoadBalancerTargetGroupsAdded(conn *autoscaling.AutoScaling, asgName string) error {
	input := &autoscaling.DescribeLoadBalancerTargetGroupsInput{
		AutoScalingGroupName: aws.String(asgName),
	}
	var tgAdding bool

	for {
		output, err := conn.DescribeLoadBalancerTargetGroups(input)

		if err != nil {
			return err
		}

		for _, tg := range output.LoadBalancerTargetGroups {
			if aws.StringValue(tg.State) == "Adding" {
				tgAdding = true
				break
			}
		}

		if tgAdding {
			tgAdding = false
			input.NextToken = nil
			continue
		}

		if aws.StringValue(output.NextToken) == "" {
			break
		}

		input.NextToken = output.NextToken
	}

	return nil
}
```

**Comment by beezly** [comment]
@bflad thanks for the review! I've added in the new test you suggested and put the waitFor functions in. The new acceptance test now passes...

`$ AWS_PROFILE=ac make testacc TEST=./aws TESTARGS='-run=TestAccAWSAutoScalingGroup_TargetGroupArns'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAutoScalingGroup_TargetGroupArns -timeout 120m
go: finding github.com/terraform-providers/terraform-provider-tls v2.1.1+incompatible
go: finding github.com/terraform-providers/terraform-provider-tls v2.1.1+incompatible
=== RUN   TestAccAWSAutoScalingGroup_TargetGroupArns
=== PAUSE TestAccAWSAutoScalingGroup_TargetGroupArns
=== CONT  TestAccAWSAutoScalingGroup_TargetGroupArns
--- PASS: TestAccAWSAutoScalingGroup_TargetGroupArns (187.36s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       187.431s`

**Comment by FernandoMiguel** [comment]
:parrot.gif:

**Pull request merged**